lib/rust/flexer/tests/flexer_generated_api_test.rs

```diff
@@ -0,0 +1,517 @@
+                FlexerStageStatus::ExitSuccess
+            }
+        }
+    }
+
+    fn gen_state_1_to_5(&mut self) -> FlexerStageStatus {
+        match u32::from(self.reader.character()) {
+            98 => FlexerStageStatus::ContinueWith(7),
+            _  => {
+                self.current_match = self.reader.pop_result();
+                self.gen_group_1_rule_1();
+                let t = self.matched_bookmark;
+                self.reader.bookmark(t);
+                FlexerStageStatus::ExitSuccess
+            }
+        }
+    }
+
+    fn gen_state_1_to_6(&mut self) -> FlexerStageStatus {
+        match u32::from(self.reader.character()) {
+            97 => FlexerStageStatus::ContinueWith(6),
+            _  => {
+                self.current_match = self.reader.pop_result();
+                self.gen_group_1_rule_0();
+                let t = self.matched_bookmark;
+                self.reader.bookmark(t);
+                FlexerStageStatus::ExitSuccess
+            }
+        }
+    }
+
+    fn gen_state_1_to_7(&mut self) -> FlexerStageStatus {
+        match u32::from(self.reader.character()) {
+            98 => FlexerStageStatus::ContinueWith(7),
+            _  => {
+                self.current_match = self.reader.pop_result();
+                self.gen_group_1_rule_1();
+                let t = self.matched_bookmark;
+                self.reader.bookmark(t);
+                FlexerStageStatus::ExitSuccess
+            }
+        }
+    }
+
+    fn gen_group_1_rule_0(&mut self) {
+        self.def_on_spaced_word_str();
+    }
+
+    fn gen_group_1_rule_1(&mut self) -> () {
+        self.def_on_spaced_word_str();
+    }
+
+    fn gen_group_1_rule_2(&mut self) {
+        self.def_on_no_err_suffix();
+    }
+
+    fn gen_group_1_rule_3(&mut self) {
+        self.def_on_err_suffix()
+    }
+}
+
+
+// === Trait Impls ===
+
+impl<Reader:LazyReader> Deref for TestLexer<Reader> {
+    type Target = Flexer<TestState,AST,Reader>;
+    fn deref(&self) -> &Self::Target {
+        &self.lexer
+    }
+}
+
+impl<Reader:LazyReader> DerefMut for TestLexer<Reader> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.lexer
+    }
+}
+
+
+
+// ===================
+// === Lexer State ===
+// ===================
+
+/// The stateful components of the test lexer.
+#[derive(Debug)]
+pub struct TestState {
+    /// The initial state of the lexer.
+    initial_state: Rc<Group>,
+    /// The state entered when the first word has been seen.
+    seen_first_word_state: Rc<Group>,
+    /// A bookmark that is set when a match occurs, allowing for rewinding if necessary.
+    matched_bookmark: BookmarkId,
+}
+
+
+// === Trait Impls ===
+
+impl <Reader:LazyReader> FlexerState<Reader> for TestState {
+    fn new(reader:&mut Reader) -> Self {
+        let initial_state         = Rc::new(Group::new(0,String::from("ROOT"),None));
+        let seen_first_word_state = Rc::new(Group::new(1,String::from("SEEN FIRST WORD"),None));
```

Comment:
this is a little bit ugly API. You should not design functions or constructors that require the user of this API to clone things explicitly. Instead, you should be able to write `Group::new(0,"ROOT",None)`. In order to do that, just accept polymorphic var `fn new(label:impl Into<String>)` - this will allow passing reference, ownership and all otherthings with native performance.